### PR TITLE
etcdbackend: support version auto discovery

### PR DESCRIPTION
**Pull request by xiang90**
Rule:

when api version is not defined by user:

1. use api 2 if server only supports api 2
2. use api 2 if previous vault data exists in api 2
3. use api 3 if server supports api3 and no user data exists in api 2

when api version is provided by user, always trust user.

Manually tested.

### CODE REVIEW

**Comment by xiang90** [comment]
@jefferai 

Hi, i do not think the CI failure is relevant to this pr.

**Comment by jefferai** [comment]
Is rule #2 because you're concerned about breakages with existing data if the API level changes?

**Comment by jefferai** [comment]
Also, the failure is not related to this. I made a copypasta screwup when fixing tests because of a sad interaction between Go bugs and `trusty` on Travis. It should be fixed if you rebase on `master`.

**Comment by xiang90** [comment]
> Is rule #2 because you're concerned about breakages with existing data if the API level changes?

Yes, exactly.

File: physical/etcd.go [hunk 1]
```diff
@@ -25,22 +28,67 @@ func newEtcdBackend(conf map[string]string, logger log.Logger) (Backend, error)
...
+		if remoteAPIVersion == "2" {
+			return nil, errors.New("etcd3 is required: etcd2 is running")
+		}
 		return newEtcd3Backend(conf, logger)
 	default:
 		return nil, EtcdVersionUnknow
 	}
 }
+
+func getEtcdAPIVersion(c client.Client) (string, error) {
```
Comment (vishalnayak) [inline physical/etcd.go]: Comments on this function as to how the version is computed might be helpful.

File: physical/etcd2.go [hunk 2]
```diff
@@ -67,6 +67,51 @@ func newEtcd2Backend(conf map[string]string, logger log.Logger) (Backend, error)
...
+	c, err := newEtcdV2Client(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	haEnabled := os.Getenv("ETCD_HA_ENABLED")
+	if haEnabled == "" {
+		haEnabled = conf["ha_enabled"]
+	}
+	haEnabledBool, _ := strconv.ParseBool(haEnabled)
```
Comment (vishalnayak) [inline physical/etcd2.go]: Can we catch the error here?

**Comment by jefferai** [comment]
Agreed on Vishal's request to catch the ParseBool error (people do put strange things in config file) but after that we'll merge it.

**Comment by xiang90** [comment]
@jefferai I will clean up today or tomorrow, and ping you guys. Thanks a lot for reviewing it.

**Comment by xiang90** [comment]
@jefferai @vishalnayak fixed. PTAL!

**Comment by jefferai** [comment]
👍

**Pull request merged**